kaldb/src/main/java/com/slack/kaldb/writer/SpanFormatter.java

```diff
@@ -13,16 +13,106 @@
 
+  // TODO: Take duration unit as input.
+  // TODO: Take a generic field mapping dictionary as input for fields.
+  public static Trace.Span toSpan(
+      Map<String, Object> jsonMsgMap,
+      String id,
+      String name,
+      String serviceName,
+      long timestamp,
+      long duration,
+      Optional<String> host,
+      Optional<String> traceId) {
+
+    Trace.Span.Builder spanBuilder = Trace.Span.newBuilder();
+
+    spanBuilder.setName(name);
+
+    String parentId =
+        (String) jsonMsgMap.getOrDefault(LogMessage.ReservedField.PARENT_ID.fieldName, "");
+    spanBuilder.setParentId(ByteString.copyFrom(parentId.getBytes()));
+
+    traceId.ifPresent(s -> spanBuilder.setTraceId(ByteString.copyFromUtf8(s)));
+    spanBuilder.setTimestamp(timestamp);
+    spanBuilder.setDuration(duration);
+
+    List<Trace.KeyValue> tags = new ArrayList<>(jsonMsgMap.size());
+    for (Map.Entry<String, Object> entry : jsonMsgMap.entrySet()) {
+      String key = entry.getKey();
+      if (MurronLogFormatter.nonTagFields.contains(key)) {
+        continue;
+      }
+      Trace.KeyValue.Builder tagBuilder = Trace.KeyValue.newBuilder();
+      tagBuilder.setKey(key);
+      if (entry.getValue() instanceof String) {
+        tagBuilder.setVType(Trace.ValueType.STRING);
+        tagBuilder.setVStr(entry.getValue().toString());
+      } else if (entry.getValue() instanceof Boolean) {
+        tagBuilder.setVType(Trace.ValueType.BOOL);
+        tagBuilder.setVBool((boolean) entry.getValue());
+      } else if (entry.getValue() instanceof Integer) {
+        tagBuilder.setVType(Trace.ValueType.INT64);
+        tagBuilder.setVInt64((int) entry.getValue());
+      } else if (entry.getValue() instanceof Float) {
+        tagBuilder.setVType(Trace.ValueType.FLOAT64);
+        tagBuilder.setVFloat64((float) entry.getValue());
+      } else if (entry.getValue() instanceof Double) {
+        tagBuilder.setVType(Trace.ValueType.FLOAT64);
+        tagBuilder.setVFloat64((double) entry.getValue());
+      } else if (entry.getValue() != null) {
+        tagBuilder.setVType(Trace.ValueType.BINARY);
+        tagBuilder.setVBinary(ByteString.copyFrom(entry.getValue().toString().getBytes()));
+      }
+      tags.add(tagBuilder.build());
+    }
+
+    // Add missing fields from murron message.
+    boolean containsHostName =
+        tags.stream()
+            .anyMatch(
+                keyValue -> keyValue.getKey().equals(LogMessage.ReservedField.HOSTNAME.fieldName));
+    if (!containsHostName) {
+      host.ifPresent(
+          s ->
+              tags.add(
+                  Trace.KeyValue.newBuilder()
+                      .setKey(LogMessage.ReservedField.HOSTNAME.fieldName)
+                      .setVType(Trace.ValueType.STRING)
+                      .setVStr(s)
+                      .build()));
+    }
+
+    spanBuilder.setId(ByteString.copyFrom(id.getBytes()));
+
+    boolean containsServiceName =
```

Comment:
previously without this check, if a `service_name` key was present, there would be two tags that end up with the same key